src/doc/trpl/unsafe.md

```diff
@@ -127,6 +111,101 @@ The latter assumption allows the compiler to optimize more
 effectively. As can be seen, actually *creating* a raw pointer is not
 unsafe, and neither is converting to an integer.
 
+## Reference and pointer aliasing
+
+Several examples can give a better comprehension of the reference and pointer aliasing rules.
+
+First of all, using `unsafe` code to incorrectly circumvent and violate the
+restrictions on references leads to undefined behaviour. For example,
+the following code creates two aliasing `&mut` references, and is therefore invalid.
+
+```
+use std::mem;
+let mut x: u8 = 1;
+
+let ref_1: &mut u8 = &mut x;
+let ref_2: &mut u8 = unsafe { mem::transmute(&mut *ref_1) };
+
+// oops, ref_1 and ref_2 point to the same piece of data (x) and are
+// both usable
+*ref_1 = 10;
+*ref_2 = 20;
+```
+
+When raw pointers come into play, the situation becomes more complex.
+Raw pointers can alias with anything, but at the same time they should respect
+the reference aliasing rules, so it's forbidden to write through a raw pointer,
+when the memory location it points to is borrowed by a safe Rust reference.
+
+```
+// This function prints 1, 2, 2 and one more number
+// The fourth number can be, for example, 2 in Release build and 3 in Debug build
+unsafe fn f(p: *mut i32, r: &mut i32) {
+    let mut val1 = *p;
+    println!("{}", val1);
+    *r = 2; // This value change should be visible through the pointer p
+    val1 = *p; // This load operation cannot be optimized out
+    println!("{}", val1);
+
+    let mut val2 = *r;
+    println!("{}", val2);
+    *p = 3; // This value change is illegal and may be invisible through the reference r
+    val2 = *r; // This load operation can be optimized out based on the assumption of uniqueness of the reference r
+    println!("{}", val2);
+}
+
+fn main() {
+    let mut val = 1i32;
+    unsafe { f(&mut val, &mut val) };
+}
+```
+
+So, raw pointers can be relatively safely used as observers, but an extreme care should be taken
+when performing write operations through them. You should be sure that no one else refers to the
+modified value by `&T` or `&mut T`.
+
+The structure `UnsafeCell` is special-cased in the compiler to allow the language to work correctly
+with internal mutability. A value of `UnsafeCell` can always be safely read through a reference, even if it is updated through something else (note, that we are talking only about single-threaded scenarios here).
```

Comment:
It is not wrong to alias `UnsafeCell` and write to it even in multi-threaded environment. Doing so will not invoke UB either, so the note about single-threaded environments is not relevant in a section about aliasing and UB.